File: py_ionization/electron_impact/production_rate.py
from __future__ import annotations

import numpy as np

from py_ionization._utils import energy_bin_width, solid_angle_from_theta_range

from .cross_section import electron_impact_cross_section


def electron_impact_rate(
    energy_eV,
    electron_def,
    *,
    species: str = "H",
    spacecraft_potential_eV=None,
    dphi: float = 2 * np.pi,
    theta_range=(-np.pi / 4, np.pi / 4),
    theta_type: str = "auto",
    min_energy_eV: float = 10.0,
):
    """Compute an electron impact ionization production rate.

    Parameters
    ----------
    energy_eV
        1D electron energy bins in eV.
    electron_def
        Electron differential energy flux in eV cm^-2 s^-1 sr^-1 eV^-1.
    species
        Neutral target species for the ionization cross section.
    spacecraft_potential_eV
        Optional spacecraft potential in eV. A 1D array is interpreted as one
        value per time sample and is subtracted from ``energy_eV``.
    dphi, theta_range, theta_type
        Angular integration geometry.
    min_energy_eV
        Corrected energies less than or equal to this value are ignored.

    Returns
    -------
    rate_s_1
        Ionization rate in s^-1 for one neutral particle.

    Formula
    -------
    ``rate = integral (DEF / E) sigma(E) dE dOmega``.
    """
    energy = np.asarray(energy_eV, dtype=float)
    if energy.ndim != 1:
        raise ValueError("energy_eV must be a 1D array.")
    if np.any(energy <= 0):
        raise ValueError("energy_eV must be positive.")

    flux = np.asarray(electron_def, dtype=float)
    scalar_time = flux.ndim == 1
    if scalar_time:
        flux = flux[None, :]
    elif flux.ndim != 2:
        raise ValueError("electron_def must be 1D or 2D.")
    if flux.shape[1] != energy.size:
        raise ValueError("electron_def must have the same number of energy bins as energy_eV.")

    n_time = flux.shape[0]
    energy_mat = np.tile(energy[None, :], (n_time, 1))
    if spacecraft_potential_eV is None:
        potential = np.zeros_like(energy_mat)
    else:
        scpot = np.asarray(spacecraft_potential_eV, dtype=float)
        potential = scpot[:, None] if scpot.ndim == 1 else scpot
    corrected_energy = energy_mat - potential
    corrected_energy = np.where(corrected_energy <= min_energy_eV, np.nan, corrected_energy)

    sigma = electron_impact_cross_section(energy, species=species)
    dE = energy_bin_width(energy)[None, :]
    domega, _ = solid_angle_from_theta_range(theta_range, dphi=dphi, theta_type=theta_type)
    differential_particle_flux = flux / corrected_energy
    differential_particle_flux = np.where(np.isfinite(corrected_energy), differential_particle_flux, np.nan)
    rate = np.nansum(differential_particle_flux * sigma[None, :] * domega * dE, axis=1)

    if scalar_time:
        return float(rate[0])
    return rate


ei_prod_rate = electron_impact_rate
production_rate = electron_impact_rate

__all__ = ["electron_impact_rate", "ei_prod_rate", "production_rate"]

File: matlab/+py_ionization/+electron_impact/production_rate.m
function rate_s_1 = production_rate(energy_eV, electron_def, varargin)
%PRODUCTION_RATE Compute an electron impact ionization production rate.
%
% rate_s_1 = py_ionization.electron_impact.production_rate(energy_eV, electron_def)
%
% Name-value options
%   Species                Neutral target species, default "H".
%   SpacecraftPotentialEV  Optional spacecraft potential in eV.
%   Dphi                   Azimuthal width in radians, default 2*pi.
%   ThetaRange             Theta limits in radians, default [-pi/4, pi/4].
%   ThetaType              "auto", "elevation", or "polar", default "auto".
%   MinEnergyEV            Corrected energies <= this value are ignored,
%                          default 10 eV.

p = inputParser;
p.addParameter("Species", "H", @(x) ischar(x) || isstring(x));
p.addParameter("SpacecraftPotentialEV", [], @(x) isnumeric(x) || isempty(x));
p.addParameter("Dphi", 2*pi, @(x) isnumeric(x) && isscalar(x));
p.addParameter("ThetaRange", [-pi/4, pi/4], @(x) isnumeric(x) && numel(x) == 2);
p.addParameter("ThetaType", "auto", @(x) ischar(x) || isstring(x));
p.addParameter("MinEnergyEV", 10.0, @(x) isnumeric(x) && isscalar(x));
p.parse(varargin{:});

energy = double(energy_eV(:).');
if any(energy <= 0)
    error("py_ionization:electron_impact:BadEnergy", ...
        "energy_eV must be positive.");
end

flux = double(electron_def);
scalar_input = isvector(flux);
if scalar_input
    flux = reshape(flux, 1, []);
elseif size(flux, 2) ~= numel(energy) && size(flux, 1) == numel(energy)
    flux = flux.';
end
if size(flux, 2) ~= numel(energy)
    error("py_ionization:electron_impact:SizeMismatch", ...
        "electron_def must have the same number of energy bins as energy_eV.");
end

n_time = size(flux, 1);
energy_mat = repmat(energy, n_time, 1);
scpot = p.Results.SpacecraftPotentialEV;
if isempty(scpot)
    potential = zeros(size(energy_mat));
elseif isscalar(scpot)
    potential = repmat(double(scpot), size(energy_mat));
elseif isvector(scpot)
    potential = repmat(double(scpot(:)), 1, numel(energy));
else
    potential = double(scpot);
end

corrected_energy = energy_mat - potential;
corrected_energy(corrected_energy <= p.Results.MinEnergyEV) = nan;

sigma = py_ionization.electron_impact.cross_section(energy, p.Results.Species);
dE = py_ionization.energy_bin_width(energy);
[domega, ~] = py_ionization.solid_angle_from_theta_range( ...
    p.Results.ThetaRange, "Dphi", p.Results.Dphi, "ThetaType", p.Results.ThetaType);

differential_particle_flux = flux ./ corrected_energy;
differential_particle_flux(~isfinite(corrected_energy)) = nan;
integrand = differential_particle_flux .* sigma .* domega .* dE;
integrand(~isfinite(integrand)) = 0;
rate_s_1 = sum(integrand, 2);

if scalar_input
    rate_s_1 = rate_s_1(1);
end
end

